pkg/clusteragent/admission/validate/kubernetesadmissionevents/kubernetesadmissionevents.go

```diff
@@ -0,0 +1,199 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+// Package kubernetesadmissionevents is a validation webhook that admit all requests and generate a Datadog Event.
+package kubernetesadmissionevents
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	admiv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/DataDog/datadog-agent/cmd/cluster-agent/admission"
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/admission/common"
+	validatecommon "github.com/DataDog/datadog-agent/pkg/clusteragent/admission/validate/common"
+	checkid "github.com/DataDog/datadog-agent/pkg/collector/check/id"
+	"github.com/DataDog/datadog-agent/pkg/metrics/event"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// Webhook is the KubernetesAdmissionEvents webhook.
+type Webhook struct {
+	name                    string
+	isEnabled               bool
+	endpoint                string
+	resources               map[string][]string
+	operations              []admissionregistrationv1.OperationType
+	matchConditions         []admissionregistrationv1.MatchCondition
+	demultiplexer           aggregator.Demultiplexer
+	supportsMatchConditions bool
+	checkid                 checkid.ID
+}
+
+// NewWebhook returns a new KubernetesAdmissionEvents webhook.
+func NewWebhook(datadogConfig config.Component, demultiplexer aggregator.Demultiplexer, supportsMatchConditions bool) *Webhook {
+	return &Webhook{
+		name:      "kubernetes_admission_events",
+		isEnabled: datadogConfig.GetBool("admission_controller.kubernetes_admission_events.enabled"),
+		endpoint:  "/kubernetes-admission-events",
+		resources: map[string][]string{
+			"apps": {
+				"deployments",
+			},
+		},
+		operations: []admissionregistrationv1.OperationType{
+			admissionregistrationv1.OperationAll,
+		},
+		// Only supported by Kubernetes 1.28+. Otherwise, filtering is done in the emitEvent() function.
+		matchConditions: []admissionregistrationv1.MatchCondition{
+			{
+				Name:       "exclude-system-users",
+				Expression: "!(request.userInfo.username.startsWith('system:'))",
```

Comment:
Curious: why do we want to exclude system user actions from the admission events? Is it for some security reasons?

Reply:
The goal of this feature is to only show human-made events as system events would create a lot of noise and are generally not interesting.